File: panda_gazebo/src/robot_joint_wrapper.cpp
// Subscribes to a single joint state command and publishes the data to individual joints

#include <ros/ros.h>
#include <sensor_msgs/JointState.h>
#include <std_msgs/Float64.h>
#include <std_msgs/Char.h>
#include <iostream>
#include <sstream>
#include <vector>
#include <map>
#include <string>

using namespace std;
bool got_joints_=false, publishers_init_=false, got_control_method_=false, notified_user_=false;
int n_joints_;
char control_method_;
map<string, ros::Publisher> joint_pub_;
vector<string> joint_names_;

vector<string> panda_joint_names {"panda_j1", "panda_j2", "panda_j3", "panda_j4", "panda_j5", "panda_j6", "panda_j7"};
vector<string> hithand_joint_names {"thumb1", "thumb2", "thumb3", "thumb4", "index1", "index2", "index3", "index4",  "middle1",  "middle2",  "middle3",  "middle4",  "ring1",  "ring2",  "ring3",  "ring4",  "little1",  "little2",  "little3",  "little4",};

// Listen to: /hithand/joint_cmd , then split this command and publish to /hithand/middle1_position_controller/command

/**
 * Check if a commanded control mode is valid
 *
 * @param control_mode The control mode
 *
 * @return true if valid, false otherwise
 */
bool controlArgValid(char control_mode)
{
  if(control_mode =='p' || control_mode =='v' || control_mode =='e')
  {
    return true;
  }
  else
  {
    ROS_ERROR_STREAM("[RobotJointWrapper] Unknown control method:" << control_mode
		     << " publish control_method as 'p' for position, "
		     << "'v' for velocity or 'e' for effort control");
    return false;
  }
}

/**
 * Callback function for joint commands. Processes joint command message and publishes
 * to underlynig joint topics.
 *
 * @param jc joint command message
 */
void jointCallback(sensor_msgs::JointState jc)
{
  if(got_control_method_)
  {
    bool command_size_correct;
    vector<double> position=jc.position;
    vector<double> velocity=jc.velocity;
    vector<double> effort=jc.effort;
    if((position.size() == n_joints_) || (velocity.size() == n_joints_) || (effort.size() == n_joints_)){
      command_size_correct = true;
    }
    if(!command_size_correct){
      ROS_ERROR_STREAM("The command had the wrong size. Size of command should be equal to number of controlled joints:" << n_joints_);
    }
    if(publishers_init_)
    {
      for(int i=0;i<n_joints_;i++)
      {
        std_msgs::Float64 joint_data;
        if(control_method_ == 'p')
        {
          joint_data.data=position[i];
        }
        else if(control_method_ == 'v')
        {
          joint_data.data=velocity[i];
        }
        else if(control_method_ == 'e')
        {
          joint_data.data=effort[i];
        }
        else
        {
          ROS_ERROR_STREAM("[RobotJointWrapper] Unknown control method:" << control_method_
                           << " publish control_method as 'p' for position,"
			   << "'v' for velocity or 'e' for effort control");
        }
        //Publish joint data
        joint_pub_[joint_names_[i]].publish(joint_data);
      }
    } else
    {
      // This is so you you don't continually yell at the user, which happens in Orocos sim
      if (!notified_user_)
      {
	      ROS_INFO("[RobotJointWrapper] Need to initialize publishers!");
	      notified_user_ = true;
      }
    }
  }
}


/**
 * Callback function for setting the control mode
 *
 * @param c_method char message defining the control mode between
 * position ('p'), velocity ('v'), and effort ('e')
 */
void controlCallback(std_msgs::Char c_method)
{
  if( ! controlArgValid(c_method.data) ) return;
  control_method_ = c_method.data;
  got_control_method_ = true;
}

/**
 * Starts up a node and initializes controllers
 *
 * robot name
 *
 * @return
 */
int main(int argc,char* argv[])
{
  bool set_robot_name=false;

  // Initialize the node
  ros::init(argc,argv,"robot_joint_wrapper");
  ros::NodeHandle n;
  ros::NodeHandle n_private("~");
  std::string listen_prefix;
  std::string publish_prefix;
  std::string jnt_cmd_topic;

  n_private.param<std::string>("listen_prefix", listen_prefix,"/panda");
  n_private.param<std::string>("publish_prefix", publish_prefix,"/panda"); // PUBLISH: /hithand/JOINT_position_controller/command
  n_private.param<std::string>("jnt_cmd_topic", jnt_cmd_topic, listen_prefix + "/joint_cmd"); // LISTEN: /hithand/joint_cmd

  listen_prefix.insert(0, 1, '/');
  publish_prefix.insert(0, 1, '/');

  //cout << "Listen prefix:" << listen_prefix;
  //cout << "\n Publish prefix:" << publish_prefix;

  // this section just assigns the hard coded panda/hithand joint names
  if(publish_prefix == "/panda_hithand"){
    joint_names_ = panda_joint_names;
    n_joints_ = joint_names_.size();
  }
  else if(publish_prefix == "/hithand"){
    joint_names_ = hithand_joint_names;
    n_joints_ = joint_names_.size();
  }
  else{
    ROS_ERROR_STREAM("Unknown publish prefix, only /panda or /hithand currently available. If using different robot or hand, change this script");
  }
  got_joints_ = true;

  // Setup subscribers
  stringstream topic_name;
  topic_name << listen_prefix << "/control_type";
  ros::Subscriber control_sub=n.subscribe(topic_name.str(), 100, controlCallback); // Change the control type with this

  ros::Subscriber joint_sub=n.subscribe(jnt_cmd_topic, 100, jointCallback); // This listens to the data on /robot/joint_cmd

  // Change the loop rate as required
  int loop_rate_int = 100;
  n_private.getParam("control_loop_rate", loop_rate_int);
  ros::Rate loop_rate(loop_rate_int);
  ROS_INFO_STREAM("[RobotJointWrapper] Control rate set to " << loop_rate_int);
  string control_method;
  n_private.param<string>("control_method", control_method, "p");
  ROS_DEBUG_STREAM("control_method input " << control_method << "\tset to " <<
                   control_method_);

  if (control_method.size() > 0 && controlArgValid(control_method[0]))
  {
    control_method_ = control_method[0];
    got_control_method_ = true;
    ROS_INFO_STREAM("[RobotJointWrapper] Initialized control method to " << control_method_);
  }

  stringstream pub_name;
  while(ros::ok())
  {
    if(got_joints_ && !publishers_init_)
    {
      // Init Publishers
      // joint_pub_.resize(n_joints_);
      for(int i = 0; i < n_joints_; i++)
      {
        pub_name.str(""); // empty buffer
        pub_name << publish_prefix << "/";
        if(control_method_ == 'p')
        {
          pub_name << joint_names_[i] << "_position_controller/command";
        }
        else if(control_method_ == 'v')
        {
          pub_name << joint_names_[i] << "_velocity_controller/command";
        }
        else if(control_method_ == 'e')
        {
          pub_name << joint_names_[i] << "_effort_controller/command";
        }
        else
        {
          ROS_ERROR_STREAM("[RobotJointWrapper] publish control_method as 'p' for position, "
		    << "'v' for velocity or 'e' for effort control");
        }
        joint_pub_[joint_names_[i]]=n.advertise<std_msgs::Float64>(pub_name.str(), 1);
      }
      publishers_init_ = true;
      ROS_INFO_STREAM("Publishers are now INITIALIZED, RECEIVED JOINT NAMES!");
    }
    ros::spinOnce();
    loop_rate.sleep();
  }
  return 0;
}

File: panda_gazebo/src/robot_joint_wrapper.py
#!/usr/bin/env python3
import rospy
from sensor_msgs.msg import JointState
from std_msgs.msg import Float64
import time

panda_joints = [
    "panda_j1", "panda_j2", "panda_j3", "panda_j4", "panda_j5", "panda_j6",
    "panda_j7"
]
hithand_joints = [
    'index0', 'index1', 'index2', 'index3', 'little0', 'little1', 'little2',
    'little3', 'middle0', 'middle1', 'middle2', 'middle3', 'ring0', 'ring1',
    'ring2', 'ring3', 'thumb0', 'thumb1', 'thumb2', 'thumb3'
]


class RobotJointWrapper:
    """ This class listens to the topic hithand/joint_cmd or panda/joint_cmd, splits the command and publishes it to the individual 
    joint control topics. On the command topic only the joint positions or efforts or velocities have to be specified, not the joint names, since the joints names are hardcoded in this script.
    """
    def __init__(self):
        rospy.init_node('robot_joint_wrapper')
        rospy.loginfo('Robot joint wrapper node started.')
        self.publishers_init_flag = False
        self.control_rate = rospy.get_param('~control_rate')
        rospy.loginfo("I heard the control rate: %d" % self.control_rate)
        self.publish_prefix = rospy.get_param('~publish_prefix')
        self.listen_prefix = rospy.get_param('~listen_prefix')
        self.command_topic = rospy.get_param('~joint_command_topic')
        self.control_method = rospy.get_param('~control_method')
        self.check_control_method()

        if self.listen_prefix == 'hithand':
            self.joint_names = hithand_joints
            self.number_of_joints = len(hithand_joints)
        elif self.listen_prefix == 'panda':
            self.joint_names = panda_joints
            self.number_of_joints = len(panda_joints)
        else:
            rospy.logerr('Unknown listen prefix: Should be panda or hithand')

        self.init_control_command_publishers()

        listened_joint_command_topic = '/' + self.listen_prefix + '/joint_cmd'
        self.joint_command_subscriber = rospy.Subscriber(
            listened_joint_command_topic,
            JointState,
            self.callback_listened_joint_command,
            queue_size=100)
        self.curr_time = None
        self.prev_time = None

    def init_control_command_publishers(self):
        """ Builds a list with the publishers for the individual joint control topics.
        """
        self.control_command_publishers = []
        for i in range(self.number_of_joints):
            topic = '/' + self.publish_prefix + '/' + self.joint_names[
                i] + '_' + self.controller_name + '/command'
            self.control_command_publishers.append(
                rospy.Publisher(topic, Float64, queue_size=1))
        self.publishers_init_flag = True

    def check_control_method(self):
        """ Verify that the specified control method is position, velocity or effort.
        """
        if (self.control_method == 'p'):
            self.controller_name = 'position_controller'
        elif (self.control_method == 'v'):
            self.controller_name = 'velocity_controller'
        elif (self.control_method != 'e'):
            self.controller_name = 'effort_controller'
        else:
            rospy.logerr(
                '[RobotJointWrapper] Unknown control method: Use p, v or e')

    def callback_listened_joint_command(self, joint_command):
        """ This function listenes to the control command on the robot/joint_cmd topic splits the command and publishes it to individual topics.
        """
        self.curr_time = time.time()
        if self.prev_time is not None:
            diff = self.curr_time - self.prev_time
            print(diff)
        self.prev_time = self.curr_time
        if self.publishers_init_flag:
            # First check that the command has the right length.
            if (len(joint_command.position) != self.number_of_joints) and (len(
                    joint_command.velocity) != self.number_of_joints) and (len(
                        joint_command.effort) != self.number_of_joints):
                rospy.logerr(
                    'The command has the wrong size. The correct size is: %d' %
                    self.number_of_joints)
            # Then choose the data from the command dependant on the control type
            for i in range(self.number_of_joints):
                joint_data = Float64()
                if self.control_method == 'p':
                    joint_data.data = joint_command.position[i]
                elif self.control_method == 'e':
                    joint_data.data = joint_command.effort[i]
                elif self.control_method == 'v':
                    joint_data.data = joint_command.velocity[i]
                else:
                    rospy.logerr(
                        'The command has no position, effort or velocity entry.'
                    )
                self.control_command_publishers[i].publish(joint_data)
        else:
            rospy.logerr(
                'The joint_wrapper publishers have not been initialized yet.')


if __name__ == '__main__':
    rjw = RobotJointWrapper()
    rate = rospy.Rate(rjw.control_rate)
    while not rospy.is_shutdown():
        rate.sleep()